File: Arrays/238. Product of Array Except Self/238. Product of Array Except Self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        ans = []
        l_mul = 1
        for n in nums:
            ans.append(l_mul)
            l_mul *= n
        
        r_mul = 1
        for i in range(len(nums) - 1, - 1, - 1):
            ans[i] *= r_mul
            r_mul *= nums[i]
        
        return ans

# Time: O(n)
# Space: O(1)

File: Arrays/238. Product of Array Except Self/238. Product of Array Except Self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function (nums) {
  let ans = [];
  let lMul = 1;
  for (const n of nums) {
    ans.push(lMul);
    lMul *= n;
  }

  let rMul = 1;
  for (let i = nums.length - 1; i > -1; i--) {
    ans[i] *= rMul;
    rMul *= nums[i];
  }

  return ans;
};

// Time: O(n)
// Space: O(1)
